File: algo/4485.cpp
#include <iostream>
#include <string.h>
#include <queue>
using namespace std;
int MAP[126][126];
int visit[126][126];
int INF = 1000000;
int dx[4] = {-1, 1, 0, 0};
int dy[4] = {0, 0, 1, -1};

int bfs(int n)
{
    memset(visit, INF, sizeof(visit));
    queue<pair<int, int>> q;
    q.push(make_pair(0, 0));
    visit[0][0] = MAP[0][0];

    while (!q.empty())
    {
        int x = q.front().first;
        int y = q.front().second;
        q.pop();
        for (int i = 0; i < 4; i++)
        {
            int mx = x + dx[i];
            int my = y + dy[i];
            if (mx >= 0 && my >= 0 && my < n && mx < n)
            {
                if (visit[mx][my] > visit[x][y] + MAP[mx][my])
                {
                    visit[mx][my] = visit[x][y] + MAP[mx][my];
                    q.push(make_pair(mx, my));
                }
            }
        }
    }

    return visit[n-1][n-1];
}

int main()
{
    int n = 0;
    int cnt = 0;
    cin >> n;
    while (n != 0)
    {
        cnt++;
        memset(MAP, 0, sizeof(MAP));
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                cin >> MAP[i][j];  

        cout << "Problem " << cnt << ": " << bfs(n) << endl;

        cin >> n;
    }
}

File: algo/4485.py
from collections import deque
import sys

INF = sys.maxsize
maps = []
dx = [-1, 1, 0, 0]
dy = [0, 0, 1, -1]
    
def bfs(maps):
    visit = [[INF for _ in range(len(maps))] for _ in range(len(maps))]
    # print(money_map)
    queue = deque([[0, 0]])
    visit[0][0] = maps[0][0]
    while queue:
        x = queue[0][0]
        y = queue[0][1]
        # cnt = queue[0][2]
        queue.popleft()

        for i in range(4):
            mx = dx[i] + x
            my = dy[i] + y
            if mx >=0 and my >=0 and mx < len(maps) and my < len(maps):
                if visit[x][y] + maps[mx][my] < visit[mx][my]:
                    visit[mx][my] = visit[x][y] + maps[mx][my]
                    queue.append([mx, my])
    # print(money_map)
    return visit[len(maps)-1][len(maps)-1]

if __name__ == "__main__":
    n = int(input())
    cnt = 0
    while n!= 0:
        cnt+= 1
        maps.clear()
        for i in range(n):
            maps.append(list(map(int, input().split())))
        print("Problem ", cnt, ": ", bfs(maps))
        n = int(input())